cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -254,49 +250,56 @@ private void checkAndUpdateJob(JobQueueDataset jobQueue, Job job) {
       readyJobs.add(job);
     }
 
-    private boolean runReadyJobs(DatasetContext context) throws Exception {
-      Iterator<Job> readyJobsIter = readyJobs.iterator();
+    private boolean runReadyJobs() throws Exception {
+      final Iterator<Job> readyJobsIter = readyJobs.iterator();
       while (readyJobsIter.hasNext() && !stopping) {
-        Job job = readyJobsIter.next();
-
-        // We should check the stored job's state (whether it actually is PENDING_LAUNCH), because
-        // the schedule could have gotten deleted in the meantime or the transaction that marked it as PENDING_LAUNCH
-        // may have failed / rolled back.
-        Job storedJob = jobQueue.getJob(job.getJobKey());
-        if (storedJob == null) {
-          readyJobsIter.remove();
-          continue;
-        }
-        if (job.isToBeDeleted()) {
-          // If this is true, that means the schedule was deleted/updated before the state transition from
-          // PENDING_CONSTRAINT to PENDING_LAUNCH is committed. We can just remove the job without launching it.
-          // If job.isToBeDeleted() is false, that means this job is in PENDING_LAUNCH state before the schedule
-          // is deleted/updated. We can still launch it.
-          readyJobsIter.remove();
-          jobQueue.deleteJob(job);
-          return true;
-        }
-
-        if (storedJob.getState() == Job.State.PENDING_LAUNCH) {
-          try {
-            taskRunner.launch(job);
-          } catch (Exception e) {
-            LOG.warn("Failed to run program {} in schedule {}. Skip running this program.",
-                     job.getSchedule().getProgramId(), job.getSchedule().getName(), e);
-            // don't delete the job, as it will be retried in some later iteration over the JobQueue
-            continue;
+        Transactions.execute(transactional, new TxCallable<Void>() {
```

Comment:
If there's a failure to launch a program, then the other ready jobs might also fail.
In any case, they'll be attempted within ~1 second (when the next iteration happens).